cxx/transition_true_value.hh

```diff
@@ -0,0 +1,105 @@
+// Copyright 2024
+// See LICENSE.txt
+
+#pragma once
+
+#include <algorithm>
+#include <random>
+#include <string>
+#include <unordered_map>
+
+#include "emissions/base.hh"
+#include "noisy_relation.hh"
+#include "relation.hh"
+
+typedef int T_item;
+typedef std::vector<T_item> T_items;
+
+// Updates the estimate of a latent value conditioned on noisy observations.
+template <typename ValueType>
+void transition_true_value(
+    std::mt19937* prng, T_items base_items, Relation<ValueType>* base_relation,
+    std::unordered_map<std::string, NoisyRelation<ValueType>*>
+        noisy_relations) {
+  // Since this method does not transition cluster assignments, we incorporate
+  // to/unincorporate from clusters themselves rather than relations. Empty
+  // clusters are not deleted, because observations will be re-incorporated to
+  // them. The number of observations in each cluster is the same before/after
+  // the call to this method.
+  base_relation->unincorporate_from_cluster(base_items);
+
+  // Unincorporate all noisy observations from clusters and store the noisy
+  // values.
+  std::unordered_map<std::string,
+                     std::unordered_map<T_items, ValueType, H_items>>
+      noisy_observations;
+  std::vector<ValueType> all_noisy_observations;
+  for (auto [name, rel] : noisy_relations) {
+    noisy_observations[name] = {};
+    for (const T_items& items : rel->base_to_noisy_items.at(base_items)) {
+      const ValueType& v = rel->get_value(items);
+      rel->unincorporate_from_cluster(items);
+      noisy_observations[name][items] = v;
+      all_noisy_observations.push_back(v);
+    }
+  }
+
+  // Compute logp_score of the noisy relations without the observations of
+  // this attribute.
+  double baseline_logp_score = 0.;
+  for (auto [name, rel] : noisy_relations) {
+    baseline_logp_score += rel->logp_score();
+  }
+
+  // Candidates for the next "true" value are the clean proposals from each
+  // cluster in each noisy relation.
+  std::vector<ValueType> true_value_candidates;
+  for (auto [name, rel] : noisy_relations) {
+    for (const auto& [i, em] : rel->emission_relation.clusters) {
+      true_value_candidates.push_back(
+          reinterpret_cast<Emission<ValueType>*>(em)->propose_clean(
+              all_noisy_observations, prng));
```

Comment:
I think this approach of using the union of the propose_clean proposals to generate "true" value candidates is a good one, but we should also check in with Alex to see if he has any other ideas that we should consider.

Reply:
I added a TODO to follow up on this.